src/Http/Routing/perf/Microbenchmarks/EndpointRoutingShortCircuitBenchmark.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.AspNetCore.Routing.ShortCircuit;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Routing;
+
+public class EndpointRoutingShortCircuitBenchmark
```

Comment:
Here's the result
|               Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|       NormalEndpoint | 665.7 ns | 5.81 ns | 4.53 ns | 1,502,118.8 | 0.0343 |     - |     - |      1 KB |
| ShortCircuitEndpoint | 772.2 ns | 9.12 ns | 8.08 ns | 1,295,049.2 | 0.0343 |     - |     - |      1 KB |

Reply:
What's NormalEndpoints benchmark without your changes?

Reply:
No it's with my changes. I will add another benchmark with previous implementation too. I assume we don't need to have two different implementation in the code base, right? I just run it in my local?

Reply:
Correct, we just want to confirm the before-and-after to make sure this doesn't regress anything.

Reply:
|               Method |     Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|       New Implementation without short circuit| 658.7 ns |  4.03 ns |  3.37 ns | 1,518,082.2 | 0.0343 |     - |     - |      1 KB |
|    Old Implementation | 639.7 ns | 10.42 ns | 12.41 ns | 1,563,187.5 | 0.0343 |     - |     - |      1 KB |
| New Implementation with short circuit| 791.3 ns |  3.24 ns |  2.87 ns | 1,263,689.7 | 0.0343 |     - |     - |      1 KB |

Reply:
It's a little ironic that short circuiting in slower 😆, but that's because we're not measuring any additional middleware in the pipeline. Consider adding the normal endpoint execution middleware after the routing one for a more accurate comparison. (optional)

Reply:
That's a huge impact, I don't expect us to merge this sort of regression. We should also measure the impact on the asp.net end to end benchmarks.

Reply:
1,518,082.2 vs 1,563,187.5? That's less than 3% on a microbenchmark. I doubt it would even show up on the end-to-end. Adding the endpoint execution middleware to this benchmark would make it more representative in all cases.

Reply:
Ah I looked at the wrong results.

>  I doubt it would even show up on the end-to-end

We don't have to guess, we can just run it.

Reply:
> New Implementation with short circuit	

Why is this so much worse?

Reply:
It's hitting the new code path that would normally not be hit until the EndpointMiddleware. If we include the EndpointMiddleware in the benchmark you shouldn't notice a difference.
https://github.com/dotnet/aspnetcore/pull/46713/files#diff-5019372fafa72e521b12fab1c28fd96919cb21c5294003b7777f5a46322cca54R113-R175
https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/EndpointMiddleware.cs#L34-L78

Reply:
Here's the result with both endpoint routing middleware and endpoint middleware together.
|               Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|       New Implementation without short circuit | 766.3 ns | 2.28 ns | 1.90 ns | 1,304,930.5 | 0.0343 |     - |     - |      1 KB |
| New Implementation with short circuit | 775.2 ns | 3.44 ns | 2.87 ns | 1,290,051.6 | 0.0343 |     - |     - |      1 KB |
|          Old Implementation | 728.6 ns | 6.77 ns | 5.28 ns | 1,372,541.8 | 0.0343 |     - |     - |      1 KB |